pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java

```diff
@@ -71,218 +113,331 @@ public TransformResultMetadata getResultMetadata() {
+        break;
+      default:
+        throw illegalState();
+    }
+  }
+
+  private void fillResultFloat(ProjectionBlock projectionBlock, int length) {
+    float[] leftFloatValues = _leftTransformFunction.transformToFloatValuesSV(projectionBlock);
+    switch (_rightStoredType) {
+      case INT:
+        fillIntResultArray(projectionBlock, leftFloatValues, length);
+        break;
+      case LONG:
+        fillLongResultArray(projectionBlock, leftFloatValues, length);
+        break;
+      case FLOAT:
+        fillFloatResultArray(projectionBlock, leftFloatValues, length);
+        break;
+      case DOUBLE:
+        fillDoubleResultArray(projectionBlock, leftFloatValues, length);
+        break;
+      case STRING:
+        fillStringResultArray(projectionBlock, leftFloatValues, length);
+        break;
+      default:
+        throw illegalState();
+    }
+  }
+
+  private void fillResultDouble(ProjectionBlock projectionBlock, int length) {
+    double[] leftDoubleValues = _leftTransformFunction.transformToDoubleValuesSV(projectionBlock);
+    switch (_rightStoredType) {
+      case INT:
+        fillIntResultArray(projectionBlock, leftDoubleValues, length);
+        break;
+      case LONG:
+        fillLongResultArray(projectionBlock, leftDoubleValues, length);
+        break;
+      case FLOAT:
+        fillFloatResultArray(projectionBlock, leftDoubleValues, length);
+        break;
+      case DOUBLE:
+        fillDoubleResultArray(projectionBlock, leftDoubleValues, length);
+        break;
+      case STRING:
+        fillStringResultArray(projectionBlock, leftDoubleValues, length);
+        break;
+      default:
+        throw illegalState();
+    }
+  }
+
+  private IllegalStateException illegalState() {
+    throw new IllegalStateException(String.format(
+        "Unsupported data type for comparison: [Left Transform Function [%s] result type is [%s], Right "
+            + "Transform Function [%s] result type is [%s]]", _leftTransformFunction.getName(), _leftStoredType,
+        _rightTransformFunction.getName(), _rightStoredType));
+  }
+
+  private void fillResultString(ProjectionBlock projectionBlock, int length) {
+    String[] leftStringValues = _leftTransformFunction.transformToStringValuesSV(projectionBlock);
+    String[] rightStringValues = _rightTransformFunction.transformToStringValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(leftStringValues[i].compareTo(rightStringValues[i]));
+    }
+  }
+
+  private void fillResultBytes(ProjectionBlock projectionBlock, int length) {
+    byte[][] leftBytesValues = _leftTransformFunction.transformToBytesValuesSV(projectionBlock);
+    byte[][] rightBytesValues = _rightTransformFunction.transformToBytesValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult((new ByteArray(leftBytesValues[i])).compareTo(new ByteArray(rightBytesValues[i])));
```

Comment:
Avoid creating the object
```suggestion
      _results[i] = getIntResult(ByteArray.compare(leftBytesValues[i], rightBytesValues[i]));
```

Reply:
This was lifted and shifted, but it makes sense to change it anyway.